quinn-udp/src/unix.rs

```diff
@@ -24,6 +24,38 @@ use super::{
     IO_ERROR_LOG_INTERVAL,
 };
 
+// Adapted from https://github.com/apple-oss-distributions/xnu/blob/main/bsd/sys/socket.h
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+#[repr(C)]
+#[allow(non_camel_case_types)]
+pub(crate) struct msghdr_x {
+    pub msg_name: *mut std::ffi::c_void,
+    pub msg_namelen: libc::socklen_t,
+    pub msg_iov: *mut libc::iovec,
+    pub msg_iovlen: std::ffi::c_int,
+    pub msg_control: *mut std::ffi::c_void,
+    pub msg_controllen: libc::socklen_t,
+    pub msg_flags: std::ffi::c_int,
+    pub msg_datalen: usize,
+}
+
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+extern "C" {
```

Comment:
Are you contributing this to upstream libc? (We can carry it until it gets released.)

Reply:
I wasn't planning on, since these are private APIs.

Reply:
Ohh, had missed that. I guess they might not take them. So given that these are documented as

> NOTE: This a private system call, the API is subject to change.

Should we make this an opt-in feature? Does private here only mean "subject to change" or also "will prevent your app from being allowed in the App Store"?

Reply:
I don't know. I do know that these calls have been there for 10+ years. I can put them behind a feature if you prefer.

Reply:
Maybe @ekinnear knows...

Reply:
> I can put them behind a feature if you prefer.

No strong opinion.

Anyone aware of any other project using these APIs across apple platforms (e.g. MacOS and iOS) and platform versions? If not, I think putting these behind a feature flag, thus allowing for some experimentation first, is the safest way to go.

We can e.g. enable the feature flag on Firefox Nightly, thus at least get to test it on a large set of MacOS machines, before always enabling it.